### PR TITLE
NNAPI: Support const values in binary ops

### PR DESCRIPTION
**Pull request by axitkhurana**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #59434

Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

### CODE REVIEW

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

File: torch/backends/_nnapi/serializer.py [hunk 1]
```diff
@@ -469,11 +472,14 @@ def get_tensor_operand_by_jitval_fixed_size(self, jitval):
...
 
-    def get_tensor_operand_or_constant(self, jitval):
+    def get_tensor_operand_or_constant(self, jitval, dim_value=DimOrder.PRESUMED_CONTIGUOUS):
         operand_id = self.jitval_operand_map.get(jitval)
         if operand_id is None:
             _, value = self.get_constant_value(jitval, "TensorType")
-            operand_id = self.add_tensor_operand_for_weight(value)
+            # assume constants are contiguous
+            if dim_value == DimOrder.CHANNELS_LAST:
+                value.permute(0, 2, 3, 1)
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:481]: This is not an in-place op.  I think you need to do `value = value.permute(0, 2, 3, 1)`.

Should probably first update the test so it fails due to this.

File: torch/backends/_nnapi/serializer.py [hunk 2]
```diff
@@ -469,11 +472,14 @@ def get_tensor_operand_by_jitval_fixed_size(self, jitval):
                 raise Exception("Flexible size is not supported for this operand.")
         return op_id, oper
 
-    def get_tensor_operand_or_constant(self, jitval):
+    def get_tensor_operand_or_constant(self, jitval, dim_value=DimOrder.PRESUMED_CONTIGUOUS):
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:475]: Why call this `dim_value` instead of `dim_order?

File: torch/backends/_nnapi/serializer.py [hunk 3]
```diff
@@ -1230,9 +1236,14 @@ def _do_add_binary(self, node, opcode, fuse_code, *, qparams=None):
         assert node.inputsAt(0).type().kind() == "TensorType"
         assert node.inputsAt(1).type().kind() == "TensorType"
 
-        # TODO: Should support constant as either operand.
-        in0_id, in0_oper = self.get_tensor_operand_by_jitval(node.inputsAt(0))
-        in1_id, in1_oper = self.get_tensor_operand_by_jitval(node.inputsAt(1))
+        if self.has_operand_for_jitval(node.inputsAt(0)):
+            in0_id, in0_oper = self.get_tensor_operand_by_jitval_fixed_size(node.inputsAt(0))
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:1240]: Does this need to be fixed size?  I think I had this in my version because I'm based on a revision where we're forcing fixed size here.

File: torch/backends/_nnapi/serializer.py [hunk 4]
```diff
@@ -534,8 +540,8 @@ def compute_operand_shape(self, op_id, dim, expr):
         self.flexible_shape_computation_lines.append(f"{flex_name(op_id, dim)} = {expr}")
 
     def transpose_to_nhwc(self, in_id, oper):
-        if oper.shape[2:] != (1, 1):
-            raise Exception("Automatic transpose only supported for H,W == 1,1")
+        if oper.dim_order != DimOrder.UNKNOWN_CONSTANT and oper.shape[2:] != (1, 1):
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:543]: Why was this change needed?

**Review by dreiss — CHANGES_REQUESTED** [review]

File: torch/backends/_nnapi/serializer.py [hunk 5]
```diff
@@ -469,11 +472,14 @@ def get_tensor_operand_by_jitval_fixed_size(self, jitval):
...
 
-    def get_tensor_operand_or_constant(self, jitval):
+    def get_tensor_operand_or_constant(self, jitval, dim_order=DimOrder.PRESUMED_CONTIGUOUS):
         operand_id = self.jitval_operand_map.get(jitval)
         if operand_id is None:
             _, value = self.get_constant_value(jitval, "TensorType")
-            operand_id = self.add_tensor_operand_for_weight(value)
+            # assume constants are contiguous
+            if dim_order == DimOrder.CHANNELS_LAST:
+                value = value.permute(0, 2, 3, 1).flatten().reshape(value.shape)
```
Comment (axitkhurana) [inline torch/backends/_nnapi/serializer.py:481]: Is there a better way to do this? If I don't flatten and reshape, broadcasting is a problem with tensor shapes mismatched

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

File: torch/backends/_nnapi/serializer.py [hunk 6]
```diff
@@ -418,6 +418,8 @@ def add_tensor_operand_for_weight(self, tensor):
             buf_num,
             offset,
             tsize))
+        if dim_order == DimOrder.CHANNELS_LAST:
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:421]: Can you just leave a comment here?  Something like

Because the NNAPI operand will have shape NHWC, we need to permute the torch tensor to the same dimension order to ensure that the data will be properly laid out after we call contiguous on it.

**Review by dreiss — APPROVED** [review]

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).
